server/src/main/java/org/opensearch/gateway/remote/ClusterStateChecksum.java

```diff
@@ -65,62 +72,147 @@ public class ClusterStateChecksum implements ToXContentFragment, Writeable {
-            nodesChecksum = checksumOut.getChecksum();
-
-            checksumOut.reset();
-            clusterState.coordinationMetadata().writeVerifiableTo(checksumOut);
-            coordinationMetadataChecksum = checksumOut.getChecksum();
-
-            // Settings create sortedMap by default, so no explicit sorting required here.
-            checksumOut.reset();
-            Settings.writeSettingsToStream(clusterState.metadata().persistentSettings(), checksumOut);
-            settingMetadataChecksum = checksumOut.getChecksum();
-
-            checksumOut.reset();
-            Settings.writeSettingsToStream(clusterState.metadata().transientSettings(), checksumOut);
-            transientSettingsMetadataChecksum = checksumOut.getChecksum();
-
-            checksumOut.reset();
-            clusterState.metadata().templatesMetadata().writeVerifiableTo(checksumOut);
-            templatesMetadataChecksum = checksumOut.getChecksum();
-
-            checksumOut.reset();
-            checksumOut.writeStringCollection(clusterState.metadata().customs().keySet());
-            customMetadataMapChecksum = checksumOut.getChecksum();
-
-            checksumOut.reset();
-            ((DiffableStringMap) clusterState.metadata().hashesOfConsistentSettings()).writeTo(checksumOut);
-            hashesOfConsistentSettingsChecksum = checksumOut.getChecksum();
-
-            checksumOut.reset();
-            checksumOut.writeMapValues(
-                clusterState.metadata().indices(),
-                (stream, value) -> value.writeVerifiableTo((BufferedChecksumStreamOutput) stream)
-            );
-            indicesChecksum = checksumOut.getChecksum();
+    public ClusterStateChecksum(ClusterState clusterState, ThreadPool threadpool) {
+        long start = threadpool.relativeTimeInNanos();
+        ExecutorService executorService = threadpool.executor(ThreadPool.Names.REMOTE_STATE_CHECKSUM);
+        CountDownLatch latch = new CountDownLatch(COMPONENT_SIZE);
 
-            checksumOut.reset();
-            clusterState.blocks().writeVerifiableTo(checksumOut);
-            blocksChecksum = checksumOut.getChecksum();
+        executeChecksumTask((stream) -> {
+            try {
+                clusterState.routingTable().writeVerifiableTo(stream);
+            } catch (IOException e) {
+                throw new RemoteStateTransferException("Failed to create checksum for routing table", e);
+            }
+            return null;
+        }, checksum -> routingTableChecksum = checksum, executorService, latch);
+
+        executeChecksumTask((stream) -> {
+            try {
+                clusterState.nodes().writeVerifiableTo(stream);
+            } catch (IOException e) {
+                throw new RemoteStateTransferException("Failed to create checksum for discovery nodes", e);
+            }
+            return null;
+        }, checksum -> nodesChecksum = checksum, executorService, latch);
+
```

Comment:
Can we take a Consumer<Stream> as ` clusterState.nodes().writeVerifiableTo(stream);` and make the entire block re-usable